File: Kotlin/Code/Kotlin-Imooc/Kotlin-CoroutineSample-v1.3/src/main/java/com/bennyhuo/coroutines/sample/Ex10_JavaInterop.java
package com.bennyhuo.coroutines.sample;

import com.bennyhuo.coroutines.utils.LogKt;

public class Ex10_JavaInterop {
    public static void main(String... args) {
        try {
            LogKt.log(CoroutineInterop.loadString());
            LogKt.log(CoroutineInterop.loadFuture().get());
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: Kotlin/Code/Kotlin-Imooc/Kotlin-CoroutineSample-v1.3/src/main/java/com/bennyhuo/coroutines/sample/Ex10_JavaInterop.kt
@file:JvmName("CoroutineInterop")
package com.bennyhuo.coroutines.sample

import kotlinx.coroutines.*
import kotlinx.coroutines.future.future

private suspend fun loadAsync(): String {
    delay(1000L)
    return "Hello"
}

fun loadString() = runBlocking {
    withContext(Dispatchers.Default) { loadAsync() }
}

//需要 org.jetbrains.kotlinx:kotlinx-coroutines-jdk8:0.22.5，Android 上面目前仍不方便使用
fun loadFuture() = GlobalScope.future {
    withContext(Dispatchers.Default) { loadAsync() }
}